server/src/main/java/org/elasticsearch/search/dfs/DfsPhase.java

```diff
@@ -176,19 +178,19 @@ private void executeKnnVectorQuery(SearchContext context) throws IOException {
         for (int i = 0; i < knnSearch.size(); i++) {
             Query knnQuery = searchExecutionContext.toQuery(knnVectorQueryBuilders.get(i)).query();
             TopScoreDocCollector topScoreDocCollector = TopScoreDocCollector.create(knnSearch.get(i).k(), Integer.MAX_VALUE);
-            Collector collector = topScoreDocCollector;
+            CollectorManager<Collector, Void> collectorManager = SingleThreadCollectorManager.wrap(topScoreDocCollector);
             if (context.getProfilers() != null) {
-                InternalProfileCollector ipc = new InternalProfileCollector(
-                    topScoreDocCollector,
+                InternalProfileCollectorManager ipcm = new InternalProfileCollectorManager(
+                    collectorManager,
                     CollectorResult.REASON_SEARCH_TOP_HITS,
                     List.of()
                 );
-                QueryProfiler knnProfiler = context.getProfilers().getDfsProfiler().addQueryProfiler(ipc);
-                collector = ipc;
+                QueryProfiler knnProfiler = context.getProfilers().getDfsProfiler().addQueryProfiler(ipcm);
+                collectorManager = ipcm;
                 // Set the current searcher profiler to gather query profiling information for gathering top K docs
                 context.searcher().setProfiler(knnProfiler);
             }
-            context.searcher().search(knnQuery, collector);
+            context.searcher().search(knnQuery, collectorManager);
```

Comment:
it looks like this was migrated only because we have a collector manager for profiling?

Reply:
Is there any reason why we should not run this concurrently? And yes, it makes like easier to use collector manager here as well.

Reply:
I don't know yet, it may make sense to migrate it but it's not a top priority. I was thinking that if we leave profiling single threaded we could probably minimize the changes needed for now. For instance is it necessary to move the profiler code to a collector manager or could we keep using a collector in there?

Reply:
> For instance is it necessary to move the profiler code to a collector manager?

I think so because of the `QueryCollectorManagerContext` abstraction. I t would be extremely awkward to implement  things differently depending on profiling,